agent.go

```diff
@@ -193,9 +193,12 @@ func (c *controller) agentSetup() error {
 		}
 	}
 
+	c.Lock()
 	if c.agent != nil {
 		close(c.agentInitDone)
+		c.agentInitDone = make(chan struct{})
```

Comment:
Thanks for spotting. Yeah agree that this could cause a problem. We may have to use `c.agentInitDone` as a tristate:
* once initdone close it and make `c.agentInitDone` nil
* if it is nil then it is already closed so waiter doesn't need to wait and simply return
* if it is not nil then cache it and wait
* Recreate the channel only on a leave